frog/responsive-images.rkt

```diff
@@ -0,0 +1,116 @@
+    (test-eq? "resize" (begin
+                         (resize-image (string->path "../example/img/480px-image.gif") 10 out)
+                         (image-width out))
+              10)
+    (delete-file out))) ; catch exn?
+
+(define/contract (get-images image-path)
+  (path? . -> . (values pair? (listof pair?)))
+  (define output-path (string->path (current-image-output-dir)))
+  (unless (directory-exists? output-path)
+    (make-directory output-path))
+  (let* ([orig-size (image-width image-path)]
+         [sizes (filter ((curry >=) orig-size) (current-image-sizes))])
+    (values (cons image-path orig-size)
+            (for/list ([width sizes])
+              (let* ([filename (file-name-from-path image-path)]
+                     [extension (get-extension filename)]
+                     [out-path
+                      (build-path
+                       output-path
+                       (replace-extension
+                        filename
+                        (string-append "-" (number->string width) "." extension)))])
+                (unless (and (file-exists? out-path)
+                             (< (file-or-directory-modify-seconds image-path)
+                                (file-or-directory-modify-seconds out-path)))
+                  (resize-image image-path width out-path))
+                (cons out-path width))))))
+
+(define default-image-idx
+  (for/or ([v  (current-image-sizes)]
+           [ix (in-naturals)])
+    (and (= v (current-image-default-size)) ix)))
+
+(define/contract (make-responsive url)
+  (url? . -> . (listof pair?))
+  (define (make-url-string path)
+    (url->string (combine-url/relative
+                  (string->url "/")
+                  (path->string path))))
+  (define (absolute->relative s) (string->path (substring (path->string s) 1)))
+  (define image-path (absolute->relative (url->path url)))
+  (define-values (orig srcset) (get-images image-path))
+  (define src (make-url-string (car (if (> (length srcset) default-image-idx)
+                                        (list-ref srcset default-image-idx)
+                                        orig))))
+  (define srcset-string
+    (string-join (map (λ (srcdef)
+                        (format "~a ~aw" (make-url-string (car srcdef)) (cdr srcdef)))
+                      srcset)
+                 ", "))
+  `([src ,src]
+    [srcset ,srcset-string]
+    ,(let ((orig-width (cdr orig)))
+       `[sizes ,(if (current-image-sizes-attr)
+                    (current-image-sizes-attr)
+                    (format "(max-width: ~apx) 100vw, ~apx"
+                            orig-width orig-width))])))
+
+(define/contract (clean-resized-images)
+  (-> any)
+  (when (directory-exists? (current-image-output-dir))
+    (fold-files (λ (path type v)
+                  (when (eq? type 'file)
+                    (delete-file* path)
+                    (prn2 "Deleted ~a" path)))
+                '() (current-image-output-dir) #f)
```

Comment:
And the next two lines, too. Probably want to `define` a local variable here for all of these?